### PR TITLE
🌱 speedup tilt starting time

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
This PR proposes an approach to speed up tilt startup time from 2m and a half to ~40 seconds by running tilt-prepare util that builds all the kustomize tasks while starting cert-menager (in parallel).

As a nice side effect, It is now possible to drop all the tilt extensions from the project; also yq is not necessary anymore.

### CODE REVIEW

File: hack/tools/tilt-prepare/main.go
Comment (vincepri) [inline hack/tools/tilt-prepare/main.go:76]:
```suggestion
	// TODO: init root dynamically
	rootPath = "/Users/fpandini/go/src/sigs.k8s.io/cluster-api"
```

This should probably be fixed before merging?

/hold

File: hack/tools/tilt-prepare/main.go
Comment (vincepri) [inline hack/tools/tilt-prepare/main.go:85]: Can we use klog like we do in the other tools?

File: hack/tools/tilt-prepare/main.go
Comment (vincepri) [inline hack/tools/tilt-prepare/main.go:93]: klog.Fatal()

File: hack/tools/tilt-prepare/main.go
Comment (vincepri) [inline hack/tools/tilt-prepare/main.go:99]: klog.Fatal()

File: hack/tools/tilt-prepare/main.go
Comment (vincepri) [inline hack/tools/tilt-prepare/main.go:103]:
```suggestion
```

File: hack/tools/tilt-prepare/main.go
Comment (vincepri) [inline hack/tools/tilt-prepare/main.go:106]: Add a comment?

File: hack/tools/tilt-prepare/main.go
Comment (vincepri) [inline hack/tools/tilt-prepare/main.go:109]:
```suggestion
	// Create a `make task` for each tool specified defined using the --tools flag.
```

File: hack/tools/tilt-prepare/main.go
Comment (vincepri) [inline hack/tools/tilt-prepare/main.go:117]: Add a comment?

File: hack/tools/tilt-prepare/main.go
Comment (vincepri) [inline hack/tools/tilt-prepare/main.go:195]: If you do have multiple errors, closing the errors channel might let those other goroutine panic

File: hack/tools/tilt-prepare/main.go
Comment (vincepri) [inline hack/tools/tilt-prepare/main.go:208]: Move this in the defer above?

File: hack/tools/tilt-prepare/main.go
Comment (vincepri) [inline hack/tools/tilt-prepare/main.go:309]:
```suggestion
		kustomize := exec.Command(kustomizePath, "build", path)
```

File: hack/tools/tilt-prepare/main.go
Comment (vincepri) [inline hack/tools/tilt-prepare/main.go:319]:
```suggestion
		envsubst := exec.Command(envsubstPath)
```

File: hack/tools/tilt-prepare/main.go
Comment (vincepri) [inline hack/tools/tilt-prepare/main.go:362]:
```suggestion
// writeIfChanged writes yaml to a file if the file does not exist or if the content has changed.
```

File: hack/tools/tilt-prepare/main.go
Comment (vincepri) [inline hack/tools/tilt-prepare/main.go:392]: Add comment?

File: hack/tools/tilt-prepare/main.go
Comment (vincepri) [inline hack/tools/tilt-prepare/main.go:415]: Should we add an `||` to check both? Or check prefix instead?

**Review by killianmuldoon — COMMENTED** [review]
It's definitely a lot quicker 😄

File: hack/tools/tilt-prepare/main.go
Comment (killianmuldoon) [reply hack/tools/tilt-prepare/main.go:76]: +1 definitely - this was the only barrier to getting this running on my machine.

Comment (fabriziopandini) [reply hack/tools/tilt-prepare/main.go:76]: 😅 
Fixed

File: hack/tools/tools.go
Comment (sbueringer) [inline hack/tools/tools.go:30]:
```suggestion
```
I don't think we have to add them here. We only have to add binaries, because if we wouldn't they would be dropped from the go.mod file. But those 2 dependencies will be part of go.mod because they are used in tilt-prepare.

**Review by sbueringer — COMMENTED** [review]
Very nice!

Only did a quick high-level review. But overall looks great. 

I would do another more detailed review after the first round of fixes.

File: hack/tools/tilt-prepare/main.go
Comment (sbueringer) [reply hack/tools/tilt-prepare/main.go:415]: Is it also possible to specify it via two separate array elements?
```yaml
    - --leader-elect
    - true
```

File: hack/tools/tilt-prepare/main.go
Comment (sbueringer) [reply hack/tools/tilt-prepare/main.go:76]: I think you didn't push it :)

File: hack/tools/tilt-prepare/main.go
Comment (fabriziopandini) [reply hack/tools/tilt-prepare/main.go]: klog.Fatal creates a panic, which gives a bad ux in tilt dashboard

File: hack/tools/tilt-prepare/main.go
Comment (fabriziopandini) [reply hack/tools/tilt-prepare/main.go]: klog.Fatal creates a panic, which gives a bad ux in tilt dashboard

File: hack/tools/tilt-prepare/main.go
Comment (fabriziopandini) [reply hack/tools/tilt-prepare/main.go:238]: it seems that moving this defer into the previous one makes the wg to finish immediately

File: hack/tools/tilt-prepare/main.go
Comment (fabriziopandini) [reply hack/tools/tilt-prepare/main.go]: I have managed the two most common use cases, then we can eventually iterate

**Comment by vincepri** [comment]
LGTM pending squash

File: hack/tools/tilt-prepare/main.go
Comment (sbueringer) [reply hack/tools/tilt-prepare/main.go:502]: Note: I'm not sure if I did something wrong, but building a new version of tilt-prepare required to run `make clean-bin` manually.

**Review by sbueringer — COMMENTED** [review]
Great step forward, thx!!

File: Tiltfile
Comment (fabriziopandini) [reply Tiltfile:93]: dropped by another PR

File: hack/tools/tilt-prepare/main.go
Comment (fabriziopandini) [reply hack/tools/tilt-prepare/main.go:241]: I don't think it is required, this is a catch-all for many tasks failing at the same time, not being properly canceled, and failing to write on an already closed channel.

File: hack/tools/tilt-prepare/main.go
Comment (fabriziopandini) [reply hack/tools/tilt-prepare/main.go:281]: +1 to create an issue

File: hack/tools/tilt-prepare/main.go
Comment (fabriziopandini) [reply hack/tools/tilt-prepare/main.go:302]: I will look into respecting allowed_contexts, but I think we should avoid having two ways for deploying cert-manager
/hold

File: hack/tools/tilt-prepare/main.go
Comment (fabriziopandini) [reply hack/tools/tilt-prepare/main.go:328]: this should be possible given that the images we are deploying in this stage are not built by tilt, but it requires some investigations. it is future work

File: hack/tools/tilt-prepare/main.go
Comment (fabriziopandini) [reply hack/tools/tilt-prepare/main.go:502]: that's how all the hack/tools are managed...

File: hack/tools/tilt-prepare/main.go
Comment (sbueringer) [reply hack/tools/tilt-prepare/main.go:302]: Yup agree. Maybe we can take the context tilt is using and use it here too. Then we're automatically covered.

_speaking as someone who has deployed on prod clusters because of tools which were using current context in the past_

File: hack/tools/tilt-prepare/main.go
Comment (sbueringer) [reply hack/tools/tilt-prepare/main.go:357]: Yup sure. Was just wondering. 

Just realized that kustomizeTasks are used for observability, while the providers where tilt builds images is using providerTask. I was mixing those two up.

File: hack/tools/tilt-prepare/main.go
Comment (fabriziopandini) [reply hack/tools/tilt-prepare/main.go:331]: fixed

**Comment by fabriziopandini** [comment]
@sbueringer nit addressed + now tilt prepare is respecting allow_k8s_context
/hold cancel

File: hack/tools/tilt-prepare/main.go
Comment (sbueringer) [reply hack/tools/tilt-prepare/main.go:331]: Just saw your solution afterwards, looks good

File: hack/tools/tilt-prepare/main.go
Comment (sbueringer) [inline hack/tools/tilt-prepare/main.go:530]: Missed that one (I cannot create a new comment at a line where there is already an old one)
```go
Name: manifestLabel
```

**Review by sbueringer — COMMENTED** [review]
Looks good overall.

Should we add a short note to the migration doc that tilt-settings are not required anymore?

File: hack/tools/tilt-prepare/main.go
Comment (sbueringer) [reply hack/tools/tilt-prepare/main.go:310]: Done > https://github.com/kubernetes-sigs/cluster-api/issues/5820

File: hack/tools/tilt-prepare/main.go
Comment (fabriziopandini) [reply hack/tools/tilt-prepare/main.go:530]: fixed

**Comment by fabriziopandini** [comment]
> Should we also add a short note to the migration doc that tilt-settings are not required anymore?

Tilt settings is stille required

**Comment by sbueringer** [comment]
> > Should we also add a short note to the migration doc that tilt-settings are not required anymore?
> 
> Tilt settings is still required

My bad, misread the Pr description.

/lgtm

**Comment by schrej** [comment]
As a note: The latest kustomize version also brings a massive performance boost. I think it was also discussed somewhere in slack to upgrade the hack/ensure-kustomize.sh script to install 4.x.

**Comment by sbueringer** [comment]
> As a note: The latest kustomize version also brings a massive performance boost (I think 20s instead of 2mins on my machine). I think it was also discussed somewhere in slack to upgrade the hack/ensure-kustomize.sh script to install 4.x.

Good to know. Corresponding issue https://github.com/kubernetes-sigs/cluster-api/issues/5570

**Review by vincepri — APPROVED** [review]
/approve

**Pull request merged**
